daemon/src/identify.rs

```diff
@@ -0,0 +1,173 @@
+    type Error = ConversionError;
+
+    fn try_from(identity_msg: protocol::IdentifyMsg) -> Result<Self, Self::Error> {
+        let identity_info = PeerInfo {
+            wire_version: identity_msg.wire_version(),
+            daemon_version: identity_msg.daemon_version()?,
+            environment: identity_msg.environment().into(),
+            protocols: identity_msg.protocols(),
+        };
+
+        Ok(identity_info)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Conversion to identity info failed: {error}")]
+pub struct ConversionError {
+    #[from]
+    error: anyhow::Error,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::identify::PeerInfo;
+    use crate::Environment;
+    use libp2p_core::PublicKey;
+    use std::collections::HashSet;
+    use std::time::Duration;
+    use tokio::sync::watch;
+    use xtra::spawn::TokioGlobalSpawnExt;
+    use xtra::Actor as _;
+    use xtra::Address;
+    use xtra::Context;
+    use xtra_libp2p::endpoint::Subscribers;
+    use xtra_libp2p::libp2p::identity::Keypair;
+    use xtra_libp2p::libp2p::multiaddr::Protocol;
+    use xtra_libp2p::libp2p::transport::MemoryTransport;
+    use xtra_libp2p::libp2p::Multiaddr;
+    use xtra_libp2p::libp2p::PeerId;
+    use xtra_libp2p::Connect;
+    use xtra_libp2p::Endpoint;
+    use xtra_libp2p::ListenOn;
+
+    #[tokio::test]
+    async fn both_parties_request_identify_info_on_connection_established() {
+        let (maker_peer_id, maker_endpoint, maker_receiver) = create_endpoint_with_identify(
+            "0.4.22".to_string(),
+            Environment::Unknown,
+            Keypair::generate_ed25519().public(),
+            HashSet::new(),
+            HashSet::from(["some_maker_protocol".to_string()]),
+        );
+        let (_, taker_endpoint, taker_receiver) = create_endpoint_with_identify(
+            "0.4.22".to_string(),
+            Environment::Umbrel,
+            Keypair::generate_ed25519().public(),
+            HashSet::new(),
+            HashSet::from(["some_taker_protocol".to_string()]),
+        );
+
+        maker_endpoint
+            .send(ListenOn(Multiaddr::empty().with(Protocol::Memory(1000))))
+            .await
+            .unwrap();
+        taker_endpoint
+            .send(Connect(
+                Multiaddr::empty()
+                    .with(Protocol::Memory(1000))
+                    .with(Protocol::P2p(maker_peer_id.into())),
+            ))
+            .await
+            .unwrap()
+            .unwrap();
+
+        let taker_to_maker_peer_info = || taker_receiver.borrow().clone();
+        let maker_peer_info = retry_until_some(taker_to_maker_peer_info).await;
+
+        let maker_to_taker_peer_info = || maker_receiver.borrow().clone();
+        let taker_peer_info = retry_until_some(maker_to_taker_peer_info).await;
+
+        let expected_maker_peer_info = PeerInfo {
+            wire_version: "0.3.0".to_string(),
+            daemon_version: "0.4.22".to_string(),
```

Comment:
would be good to havea test with some of the non-released `daemon-versions`. should we strip vergen version to bare last-released-version for identify purposes?

I'll write up a ticket in a sec detailing this behaviour. I believe it can be done in a follow-up though, no need to wait with merging.

Reply:
please link the ticket if you can :)